<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>rainbow (code) â€¢ spliceQTL</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="rainbow (code)">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">spliceQTL</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/code.html">rainbow (code)</a>
    </li>
    <li>
      <a href="../articles/text.html">rainbow (text)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rauschenberger/spliceQTL">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>rainbow (code)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/rauschenberger/spliceQTL/blob/master/vignettes/code.Rmd"><code>vignettes/code.Rmd</code></a></small>
      <div class="hidden name"><code>code.Rmd</code></div>

    </div>

    
    
<p>This vignette requires both a local machine and the BBMRI virtual machine. On the virtual machine, execute this chunk to set the library path, update the R package spliceQTL, and set the working directory.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lib &lt;-<span class="st"> "/virdir/Scratch/arauschenberger/library"</span>
<span class="kw">.libPaths</span>(lib)
devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"rauschenberger/spliceQTL"</span>,<span class="dt">lib=</span>lib)
<span class="kw">library</span>(<span class="st">"spliceQTL"</span>,<span class="dt">lib.loc=</span>lib)
path &lt;-<span class="st"> "/virdir/Scratch/arauschenberger/spliceQTL"</span>
<span class="kw">setwd</span>(path)</code></pre></div>
<p>(Please always execute this chunk when starting a new R session on the virtual machine!)</p>
<div id="prepare-data" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-data" class="anchor"></a>Prepare data</h1>
<p>On a local machine with PLINK, execute this chunk to obtain the Geuvadis SNP data. Then move the files from the local to the virtual machine.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(chr <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>){
    spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/get.snps.geuvadis">get.snps.geuvadis</a></span>(<span class="dt">chr=</span>chr,<span class="dt">data=</span><span class="st">"N:/semisup/data/eQTLs"</span>,
                                 <span class="dt">path=</span><span class="st">"N:/spliceQTL/data/Geuvadis"</span>)
}</code></pre></div>
<p>On the virtual machine, execute this chunk to obtain the BBMRI SNP data, the Geuvadis exon data, and the BBMRI exon data. Choose one out of the six biobanks (CODAM, LL, LLS, NTR, PAN, RS).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(chr <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>){
    spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/get.snps.bbmri">get.snps.bbmri</a></span>(<span class="dt">chr=</span>chr,<span class="dt">biobank=</span><span class="st">"LLS"</span>,<span class="dt">path=</span>path,<span class="dt">size=</span><span class="dv">500</span><span class="op">*</span><span class="dv">10</span><span class="op">^</span><span class="dv">3</span>)
}
spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/get.exons.geuvadis">get.exons.geuvadis</a></span>(<span class="dt">path=</span>path)
spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/get.exons.bbmri">get.exons.bbmri</a></span>(<span class="dt">path=</span>path)</code></pre></div>
<p>On the virtual machine, execute this chunk to prepare the data. See the documentation of the R package spliceQTL for further information. (It seems that lme4::lmer in spliceQTL::adjust.variables fails to release memory. Restart R after each chromosome.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">for</span>(chr <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>){
  <span class="cf">for</span>(data <span class="cf">in</span> <span class="kw">c</span>(<span class="st">"Geuvadis"</span>,<span class="st">"LLS"</span>)){
      
    <span class="kw">rm</span>(<span class="dt">list=</span><span class="kw">setdiff</span>(<span class="kw">ls</span>(),<span class="kw">c</span>(<span class="st">"data"</span>,<span class="st">"chr"</span>,<span class="st">"path"</span>))); <span class="kw">gc</span>()
    <span class="kw">set.seed</span>(<span class="dv">1</span>)
    
    <span class="kw">cat</span>(<span class="st">"Analysing"</span>,data,chr,<span class="st">":"</span>,<span class="kw">as.character</span>(<span class="kw">Sys.time</span>()),<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    <span class="cf">if</span>(data<span class="op">==</span><span class="st">"Geuvadis"</span>){
      <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">"Geuvadis.exons.RData"</span>),<span class="dt">verbose=</span><span class="ot">TRUE</span>)
    } <span class="cf">else</span> {
      <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="st">"BBMRI.exons.RData"</span>),<span class="dt">verbose=</span><span class="ot">TRUE</span>)
      cond &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="dt">x=</span><span class="kw">rownames</span>(exons),<span class="dt">split=</span><span class="st">":"</span>),<span class="cf">function</span>(x) x[[<span class="dv">1</span>]]<span class="op">==</span>data)
      exons &lt;-<span class="st"> </span>exons[cond,]
    }
    <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="kw">paste0</span>(data,<span class="st">".chr"</span>,chr,<span class="st">".RData"</span>)),<span class="dt">verbose=</span><span class="ot">TRUE</span>)
    
    <span class="kw">cat</span>(<span class="st">"Matching samples:"</span>,<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    list &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/match.samples">match.samples</a></span>(exons,snps)
    exons &lt;-<span class="st"> </span>list<span class="op">$</span>exons; snps &lt;-<span class="st"> </span>list<span class="op">$</span>snps; <span class="kw">rm</span>(list)
    
    <span class="kw">cat</span>(<span class="st">"Adjusting samples:"</span>,<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    exons &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/adjust.samples">adjust.samples</a></span>(<span class="dt">x=</span>exons) <span class="co"># slow!</span>
    exons &lt;-<span class="st"> </span><span class="kw">asinh</span>(<span class="dt">x=</span>exons)
    
    <span class="kw">cat</span>(<span class="st">"Adjusting covariates:"</span>,<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    names &lt;-<span class="st"> </span><span class="kw">strsplit</span>(<span class="dt">x=</span><span class="kw">colnames</span>(exons),<span class="dt">split=</span><span class="st">"_"</span>) <span class="co"># exon names</span>
    length &lt;-<span class="st"> </span><span class="kw">sapply</span>(names,<span class="cf">function</span>(x) <span class="kw">as.integer</span>(x[[<span class="dv">3</span>]])<span class="op">-</span><span class="kw">as.integer</span>(x[[<span class="dv">2</span>]])) <span class="co"># exon length</span>
    exons &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/adjust.variables">adjust.variables</a></span>(<span class="dt">x=</span>exons,<span class="dt">group=</span>gene_id,<span class="dt">offset=</span>length) <span class="co"># slow!</span>
    
    <span class="co"># subset chromosome</span>
    cond &lt;-<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="dt">x=</span><span class="kw">colnames</span>(exons),<span class="dt">split=</span><span class="st">"_"</span>),<span class="cf">function</span>(x) x[[<span class="dv">1</span>]]<span class="op">==</span>chr)
    exons &lt;-<span class="st"> </span>exons[,cond]
    gene_id &lt;-<span class="st"> </span>gene_id[cond]
    
    <span class="kw">cat</span>(<span class="st">"Mapping exons:"</span>,<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    map &lt;-<span class="st"> </span><span class="kw">list</span>()
    map<span class="op">$</span>genes &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/map.genes">map.genes</a></span>(<span class="dt">chr=</span>chr,<span class="dt">path=</span>path)
    map<span class="op">$</span>exons &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/map.exons">map.exons</a></span>(<span class="dt">gene=</span><span class="kw">as.character</span>(map<span class="op">$</span>genes<span class="op">$</span>gene_id),<span class="dt">exon=</span>gene_id)
    
    <span class="kw">cat</span>(<span class="st">"Mapping SNPs:"</span>,<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    names &lt;-<span class="st"> </span><span class="kw">strsplit</span>(<span class="dt">x=</span><span class="kw">colnames</span>(snps),<span class="dt">split=</span><span class="st">":"</span>)
    snp.chr &lt;-<span class="st"> </span><span class="kw">sapply</span>(names,<span class="cf">function</span>(x) <span class="kw">as.integer</span>(x[[<span class="dv">1</span>]]))
    snp.pos &lt;-<span class="st"> </span><span class="kw">sapply</span>(names,<span class="cf">function</span>(x) <span class="kw">as.integer</span>(x[[<span class="dv">2</span>]]))
    map<span class="op">$</span>snps &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/map.snps">map.snps</a></span>(<span class="dt">gene.chr=</span>map<span class="op">$</span>genes<span class="op">$</span>chr,
                                    <span class="dt">gene.start=</span>map<span class="op">$</span>genes<span class="op">$</span>start,
                                    <span class="dt">gene.end=</span>map<span class="op">$</span>genes<span class="op">$</span>end,
                                    <span class="dt">snp.chr=</span>snp.chr,<span class="dt">snp.pos=</span>snp.pos)
    
    <span class="kw">cat</span>(<span class="st">"Dropping genes:"</span>,<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    map &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/drop.trivial">drop.trivial</a></span>(<span class="dt">map=</span>map)
    
    <span class="kw">cat</span>(<span class="st">"Testing:"</span>,<span class="kw">as.character</span>(<span class="kw">Sys.time</span>()),<span class="st">" -&gt; "</span>)
    <span class="kw">rm</span>(<span class="dt">list=</span><span class="kw">setdiff</span>(<span class="kw">ls</span>(),<span class="kw">c</span>(<span class="st">"exons"</span>,<span class="st">"snps"</span>,<span class="st">"map"</span>,<span class="st">"data"</span>,<span class="st">"chr"</span>,<span class="st">"path"</span>))); <span class="kw">gc</span>()
    
    <span class="kw">save</span>(<span class="dt">list=</span><span class="kw">c</span>(<span class="st">"exons"</span>,<span class="st">"snps"</span>,<span class="st">"map"</span>),<span class="dt">file=</span><span class="kw">file.path</span>(path,<span class="kw">paste0</span>(<span class="st">"temp."</span>,data,<span class="st">".chr"</span>,chr,<span class="st">".RData"</span>)))
  }
<span class="co">#q()</span>
<span class="co">#n</span>
<span class="co">#exit</span>
}</code></pre></div>
</div>
<div id="test-hypotheses" class="section level1">
<h1 class="hasAnchor">
<a href="#test-hypotheses" class="anchor"></a>Test hypotheses</h1>
<p>On the virtual machine, execute this chunk to test for alternative splicing.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">trial &lt;-<span class="st"> "/virdir/Scratch/arauschenbe/spliceQTL/trial"</span>
<span class="cf">for</span>(chr <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>){
    <span class="cf">for</span>(data <span class="cf">in</span> <span class="kw">c</span>(<span class="st">"Geuvadis"</span>,<span class="st">"LLS"</span>)){
        <span class="kw">cat</span>(<span class="st">"Analysing"</span>,data,chr,<span class="st">":"</span>,<span class="kw">as.character</span>(<span class="kw">Sys.time</span>()),<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
        <span class="kw">rm</span>(<span class="dt">list=</span><span class="kw">setdiff</span>(<span class="kw">ls</span>(),<span class="kw">c</span>(<span class="st">"data"</span>,<span class="st">"chr"</span>,<span class="st">"path"</span>))); <span class="kw">gc</span>(); <span class="kw">cat</span>(<span class="st">"."</span>)
        <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="kw">paste0</span>(<span class="st">"temp."</span>,data,<span class="st">".chr"</span>,chr,<span class="st">".RData"</span>))); <span class="kw">cat</span>(<span class="st">"."</span>)
        <span class="cf">if</span>(<span class="ot">FALSE</span>){ <span class="co"># only for development</span>
            exons &lt;-<span class="st"> </span>exons[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,]
            snps &lt;-<span class="st"> </span>snps[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,]
            map<span class="op">$</span>genes &lt;-<span class="st"> </span>map<span class="op">$</span>genes[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,]
            map<span class="op">$</span>exons &lt;-<span class="st"> </span>map<span class="op">$</span>exons[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]
            map<span class="op">$</span>snps &lt;-<span class="st"> </span>map<span class="op">$</span>snps[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,]
        }
        pvalue &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/test.multiple">test.multiple</a></span>(<span class="dt">Y=</span>exons, <span class="dt">X=</span>snps, <span class="dt">map=</span>map,<span class="dt">spec=</span><span class="dv">16</span>,<span class="dt">w.type=</span><span class="st">"cov"</span>); <span class="kw">cat</span>(<span class="st">"."</span>)
        <span class="kw">save</span>(<span class="dt">object=</span>pvalue,<span class="dt">file=</span><span class="kw">file.path</span>(trial,<span class="kw">paste0</span>(<span class="st">"pval."</span>,data,<span class="st">".chr"</span>,chr,<span class="st">".RData"</span>))); <span class="kw">cat</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)
    }
}</code></pre></div>
</div>
<div id="analyse-results" class="section level1">
<h1 class="hasAnchor">
<a href="#analyse-results" class="anchor"></a>Analyse results</h1>
<p>On the virtual machine, execute this chunk to compare the results between the Geuvadis and the BBMRI project.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">table &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">chr=</span><span class="kw">seq_len</span>(<span class="dv">22</span>))
list &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>){
  <span class="co"># load p-values</span>
  <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="kw">paste0</span>(<span class="st">"pval.Geuvadis.chr"</span>,i,<span class="st">".RData"</span>)))
  a &lt;-<span class="st"> </span>pvalue; pvalue &lt;-<span class="st"> </span><span class="ot">NA</span>
  <span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="kw">paste0</span>(<span class="st">"pval.LLS.chr"</span>,i,<span class="st">".RData"</span>)))
  b &lt;-<span class="st"> </span>pvalue; pvalue &lt;-<span class="st"> </span><span class="ot">NA</span>
  
  <span class="co"># filter tests (original)</span>
  <span class="co">#sel &lt;- "rho=1"</span>
  <span class="co">#names &lt;- intersect(rownames(a),rownames(b))</span>
  <span class="co">#a &lt;- a[names,sel]</span>
  <span class="co">#b &lt;- b[names,sel]</span>
  
  <span class="co"># filter tests (new)</span>
  names &lt;-<span class="st"> </span><span class="kw">intersect</span>(<span class="kw">rownames</span>(a),<span class="kw">rownames</span>(b))
  a &lt;-<span class="st"> </span>a[names]
  b &lt;-<span class="st"> </span>b[names]
  
  <span class="cf">if</span>(<span class="ot">FALSE</span>){
   <span class="co"># "Manhattan" plot</span>
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/frame">plot.new</a></span>()
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/plot.window">plot.window</a></span>(<span class="dt">xlim=</span><span class="kw">c</span>(<span class="fl">0.5</span>,<span class="kw">length</span>(a)<span class="op">+</span><span class="fl">0.5</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>,<span class="dv">1</span>)<span class="op">*</span><span class="kw">max</span>(<span class="op">-</span><span class="kw">log10</span>(<span class="kw">c</span>(a,b))))
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/box">box</a></span>()
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/axis">axis</a></span>(<span class="dt">side=</span><span class="dv">1</span>)
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/axis">axis</a></span>(<span class="dt">side=</span><span class="dv">2</span>)
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/segments">segments</a></span>(<span class="dt">x0=</span><span class="kw">seq_along</span>(a),<span class="dt">y0=</span><span class="dv">0</span>,<span class="dt">y1=</span><span class="op">-</span><span class="kw">log10</span>(a),<span class="dt">col=</span><span class="st">"darkred"</span>)
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/segments">segments</a></span>(<span class="dt">x0=</span><span class="kw">seq_along</span>(b),<span class="dt">y0=</span><span class="dv">0</span>,<span class="dt">y1=</span><span class="kw">log10</span>(b),<span class="dt">col=</span><span class="st">"darkblue"</span>)
   graphics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/graphics/topics/abline">abline</a></span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">"grey"</span>)
  }
  
  <span class="co"># correlation</span>
  <span class="co">#file &lt;- file.path(path,"plots",paste0("cor.chr",i,".pdf")) # activate!</span>
  <span class="co">#grDevices::pdf(file=file,width=4,height=4) # activate!</span>
  <span class="co">#spliceQTL::grid(x=-log(a),y=-log(b),n=20) # activate!</span>
  <span class="co">#grDevices::dev.off() # activate!</span>
  table<span class="op">$</span>cor[i] &lt;-<span class="st"> </span><span class="kw">round</span>(stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/cor">cor</a></span>(<span class="op">-</span><span class="kw">log</span>(a),<span class="op">-</span><span class="kw">log</span>(b),<span class="dt">method=</span><span class="st">"pearson"</span>),<span class="dt">digits=</span><span class="dv">2</span>)
  table<span class="op">$</span>cor.test[i] &lt;-<span class="st"> </span><span class="kw">signif</span>(stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/cor.test">cor.test</a></span>(<span class="op">-</span><span class="kw">log</span>(a),<span class="op">-</span><span class="kw">log</span>(b),<span class="dt">method=</span><span class="st">"pearson"</span>)<span class="op">$</span>p.value,<span class="dt">digits=</span><span class="dv">2</span>)
  
  <span class="co"># significance</span>
  a &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/p.adjust">p.adjust</a></span>(a)<span class="op">&lt;</span><span class="fl">0.05</span>
  b &lt;-<span class="st"> </span>stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/p.adjust">p.adjust</a></span>(b)<span class="op">&lt;</span><span class="fl">0.05</span>
  table<span class="op">$</span>none[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="op">!</span>a <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>b)
  table<span class="op">$</span>onlyGEU[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(a <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>b)
  table<span class="op">$</span>onlyLLS[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="op">!</span>a <span class="op">&amp;</span><span class="st"> </span>b)
  table<span class="op">$</span>both[i] &lt;-<span class="st"> </span><span class="kw">sum</span>(a <span class="op">&amp;</span><span class="st"> </span>b)
  table<span class="op">$</span>chisq.test[i] &lt;-<span class="st"> </span><span class="kw">suppressWarnings</span>(<span class="kw">signif</span>(stats<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/stats/topics/chisq.test">chisq.test</a></span>(<span class="kw">table</span>(a,b))<span class="op">$</span>p.value,<span class="dt">digits=</span><span class="dv">2</span>))
  
  list[[i]] &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">chr=</span>i,<span class="dt">gene=</span><span class="kw">names</span>(a),<span class="dt">Geuvadis=</span><span class="dv">1</span><span class="op">*</span>a,<span class="dt">LLS=</span><span class="dv">1</span><span class="op">*</span>b,<span class="dt">row.names=</span><span class="kw">seq_along</span>(a))
  
}
<span class="kw">save</span>(table,<span class="dt">file=</span><span class="st">"table.RData"</span>)

list &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="dt">what=</span><span class="st">"rbind"</span>,<span class="dt">args=</span>list)
names &lt;-<span class="st"> </span>spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/map.genes">map.genes</a></span>(<span class="dt">chr=</span><span class="ot">NULL</span>)
list<span class="op">$</span>symbol &lt;-<span class="st"> </span>names<span class="op">$</span>gene_name[<span class="kw">match</span>(<span class="dt">x=</span>list<span class="op">$</span>gene,names<span class="op">$</span>gene_id)]
<span class="kw">save</span>(list,<span class="dt">file=</span><span class="st">"list.RData"</span>)</code></pre></div>
<p>On the virtual machine, execute this chunk to plot correlations between SNPs and exons.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">data &lt;-<span class="st"> "LLS"</span> <span class="co"># "Geuvadis" or "LLS"</span>
chr &lt;-<span class="st"> </span><span class="dv">1</span> <span class="co"># 1,2,3,...,22</span>
<span class="kw">load</span>(<span class="kw">file.path</span>(path,<span class="kw">paste0</span>(<span class="st">"temp."</span>,data,<span class="st">".chr"</span>,chr,<span class="st">".RData"</span>)))
<span class="co"># i &lt;- which(map$genes$gene_id=="...") # gene name (e.g. ENSG00000172967)</span>
spliceQTL<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/spliceQTL/topics/visualise">visualise</a></span>(<span class="dt">Y=</span>exons,<span class="dt">X=</span>snps,<span class="dt">map=</span>map,<span class="dt">i=</span>i)</code></pre></div>
<!--
xtable::print.xtable(xtable::xtable(x=table[,-1],digits=c(1,2,-2,0,0,0,0,-2)))

#wait <- TRUE
#while(wait){
#    wait <- !file.exists(...)
#    if(wait){Sys.sleep(60);cat(".")}
#}
memory and CPU usage Linux: htop
-->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#prepare-data">Prepare data</a></li>
      <li><a href="#test-hypotheses">Test hypotheses</a></li>
      <li><a href="#analyse-results">Analyse results</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Armin Rauschenberger, Renee Menezes.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
